### PR TITLE
AJAX with Django API endpoints

### PR DESCRIPTION
**Pull request by saadyousafarbi**
We will be making use of the previously developed API endpoint of ```edit_profile``` and do the following:

* User would be able to edit profile by providing new profile information,
* On saving changes, user will be displayed updated profile information on the same page (```AJAX```).

### CODE REVIEW

File: autoshop/api/v1/views.py [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 from rest_framework.response import Response
 from rest_framework.viewsets import ModelViewSet
+from rest_framework.permissions import AllowAny
 
 from autoshop.api.v1.serializers import UserProfileSerializer
 from core.models import Profile
 
 
 class UserProfileViewSet(ModelViewSet):
+    permissions_classes = (AllowAny,)
```
Comment (zubair-arbi) [inline autoshop/api/v1/views.py]: we only want authenticated users to view/edit their profiles

File: core/views.py [hunk 2]
```diff
@@ -134,26 +135,27 @@ def edit_profile(request):
...
-            request.user.username, form.errors
-        )
-        messages.error(request, 'Form is invalid' + form.errors)
-        return render(request, 'profile_edit.html', context)
+    # if request.method == 'POST':
+    #     return None
+        # post_data = request.POST.copy()
+        # post_data['date_of_birth'] = parse(post_data['date_of_birth'])
+        # form = EditProfileForm(post_data, instance=user_profile)
+        # if form.is_valid():
```
Comment (zubair-arbi) [inline core/views.py]: revert

File: templates/profile_edit.html [hunk 3]
```diff
@@ -30,3 +32,47 @@
     <button type="submit" value="Submit" class="submit"> Save Changes</button>
 </form>
 {% endblock content %}
+
+
+{% block javascript %}
+  <script>
+    $("#edit_profile").submit(function(event) {
+        event.preventDefault();
+        user_id = {{ user_info_dict.user_id }};
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: define `data` var here and add parameters in it:
```javascript
data: {
  'pk': '{{ user_info_dict.user_id }}',
}
```

File: templates/profile_edit.html [hunk 4]
```diff
@@ -30,3 +32,47 @@
...
+        bio = $("#id_bio").val();
+        gender = $("#id_gender").val();
+        date_of_birth = $("#id_date_of_birth").val();
+        mobile_number = $("#id_mobile_number").val();
+        address = $("#id_address").val();
+        city = $("#id_city").val();
+        country = $("#id_country").val();
+        endpoint_url = '/profiles/'+user_id+'/';
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: which endpoint? better name `profile_edit_api_url`

File: templates/profile_edit.html [hunk 5]
```diff
@@ -30,3 +32,47 @@
...
+                'mobile_number': mobile_number,
+                'address': address,
+                'city': city,
+                'country': country,
+            },
+            dataType: 'json',
+            success: function (data) {
+                alert("Profile updated.");
+            },
+            error :function (exception) {
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: nit: spacing

File: templates/profile_edit.html [hunk 6]
```diff
@@ -30,3 +32,47 @@
...
+                alert("Unable to update."+exception);
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: spaces around operators

**Review by zubair-arbi — CHANGES_REQUESTED** [review]

File: templates/profile_edit.html (hunk 3 above)
Comment (saadyousafarbi) [reply templates/profile_edit.html]: that is already defined within AJAX function.

Comment (zubair-arbi) [reply templates/profile_edit.html]: so define it here and just use it as assignment inside the ajax function.

File: templates/profile_edit.html [hunk 7]
```diff
@@ -30,3 +32,47 @@
...
+        event.preventDefault();
+        user_id = {{ user_info_dict.user_id }};
+        bio = $("#id_bio").val();
+        gender = $("#id_gender").val();
+        date_of_birth = $("#id_date_of_birth").val();
+        mobile_number = $("#id_mobile_number").val();
+        address = $("#id_address").val();
+        city = $("#id_city").val();
+        country = $("#id_country").val();
+        profile_edit_api_url = '/profiles/'+user_id+'/';
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: define this url at top of vars

File: templates/profile_edit.html [hunk 8]
```diff
@@ -30,3 +32,47 @@
...
+                'address': address,
+                'city': city,
+                'country': country,
+            },
+            dataType: 'json',
+            success: function (data) {
+                alert("Profile updated.");
+            },
+            error : function (exception) {
+                alert("Unable to update." + exception);
```
Comment (zubair-arbi) [inline templates/profile_edit.html]:
```javascript
alert("There was an error updating the profile:  " + exception);
```

File: autoshop/settings.py [hunk 9]
```diff
@@ -111,6 +111,11 @@
     },
 ]
 
+REST_FRAMEWORK = {
+    'DEFAULT_PERMISSION_CLASSES': (
+        'rest_framework.permissions.AllowAny',
```
Comment (zubair-arbi) [inline autoshop/settings.py]: default permission should be `IsAuthenticated`

File: templates/profile_edit.html [hunk 10]
```diff
@@ -30,3 +32,41 @@
...
+        user_id = {{ user_info_dict.user_id }};
+        data: {
+            'pk': '{{ user_info_dict.user_id }}',
+            'user': '{{ user_info_dict.user_id }}',
+            'csrfmiddlewaretoken': '{{ csrf_token }}',
+            'bio' = $("#id_bio").val();
+            'gender' = $("#id_gender").val();
+            'date_of_birth' = $("#id_date_of_birth").val();
+            'mobile_number' = $("#id_mobile_number").val();
+            'address' = $("#id_address").val();
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: be persistent with single quotes in `js`

File: templates/profile_edit.html [hunk 11]
```diff
@@ -30,3 +32,41 @@
     <button type="submit" value="Submit" class="submit"> Save Changes</button>
 </form>
 {% endblock content %}
+
+
+{% block javascript %}
+  <script>
+    $("#edit_profile").submit(function(event) {
+        event.preventDefault();
+        profile_edit_api_url = '/profiles/'+user_id+'/';
```
Comment (zubair-arbi) [inline templates/profile_edit.html]: spaces around operators

**Review by zubair-arbi — APPROVED** [review]
Just few small suggestions otherwise 👍 
Rebase and squash before merge.

**Comment by saadyousafarbi** [comment]
Squashed 14e0df8a5b7f961aa4b87d3e26251285e234a6c9

**Pull request merged**